dart/lib/src/http_client/failed_request_client.dart

```diff
@@ -0,0 +1,174 @@
+///
+/// Configured with default values, this captures requests which throw an
+/// exception.
+/// This can be for example for the following reasons:
+/// - In an browser environment this can be requests which fail because of CORS.
+/// - In an mobile or desktop application this can be requests which failed
+///   because the connection was interrupted.
+///
+/// Additionally you can configure specific HTTP response codes to be considered
+/// as a failed request. In the following example, the status codes 404 and 500
+/// are considered a failed request.
+///
+/// ```dart
+/// import 'package:sentry/sentry.dart';
+///
+/// var client = FailedRequestClient(
+///   failedRequestStatusCodes: [404, 500]
+/// );
+/// ```
+///
+/// Remarks:
+/// If this client is used as a wrapper, a call to close also closes the
+/// given client.
+///
+/// The `FailedRequestClient` can be used as a standalone client like this:
+/// ```dart
+/// import 'package:sentry/sentry.dart';
+///
+/// var client = FailedRequestClient();
+/// try {
+///  var uriResponse = await client.post('https://example.com/whatsit/create',
+///      body: {'name': 'doodle', 'color': 'blue'});
+///  print(await client.get(uriResponse.bodyFields['uri']));
+/// } finally {
+///  client.close();
+/// }
+/// ```
+///
+/// The `FailedRequestClient` can also be used as a wrapper for your own
+/// HTTP [Client](https://pub.dev/documentation/http/latest/http/Client-class.html):
+/// ```dart
+/// import 'package:sentry/sentry.dart';
+/// import 'package:http/http.dart' as http;
+///
+/// final myClient = http.Client();
+///
+/// var client = FailedRequestClient(client: myClient);
+/// try {
+///  var uriResponse = await client.post('https://example.com/whatsit/create',
+///      body: {'name': 'doodle', 'color': 'blue'});
+///  print(await client.get(uriResponse.bodyFields['uri']));
+/// } finally {
+///  client.close();
+/// }
+/// ```
+class FailedRequestClient extends BaseClient {
+  FailedRequestClient({
+    this.maxRequestBodySize = MaxRequestBodySize.small,
+    this.failedRequestStatusCodes = const [],
+    Client? client,
+    Hub? hub,
+  })  : _hub = hub ?? HubAdapter(),
+        _client = client ?? Client();
+
+  final Client _client;
+  final Hub _hub;
+
+  /// Configures up to which size request bodies should be included in events
+  final MaxRequestBodySize maxRequestBodySize;
+
+  /// Describes which HTTP status codes should be considered as a failed
+  /// requests.
+  ///
+  /// Per default no status code is considered a failed request.
+  final List<int> failedRequestStatusCodes;
+
+  @override
+  Future<StreamedResponse> send(BaseRequest request) async {
+    try {
+      final response = await _client.send(request);
+
+      if (failedRequestStatusCodes.contains(response.statusCode)) {
+        await _captureException(request: request);
+      }
+
+      return response;
+    } catch (exception, stackTrace) {
+      await _captureException(
```

Comment:
you call `_captureException` twice, as you can use optional params, maybe we could do it in the finally block only once

Reply:
No not really. The first call is for non-throwing "bad" events like 500 status codes. The second call is for failing throwing requests. 

Also I don't have access to the exception and it's stacktrace in the finally block so I would need to save them to variables. For bad requests I would also need to save it to a variable. How would that be better?

Reply:
I'd not say its better, its just a conventional if you want to do something within the try and catch block, see https://github.com/getsentry/sentry-java/blob/main/sentry-android-okhttp/src/main/java/io/sentry/android/okhttp/SentryOkHttpInterceptor.kt

also, let's say you have a bug in the `_captureException` method that throws if the request is broken, it throws and go to `catch` block, it calls `_captureException` again, its a side effect.

Reply:
Ah that's a good point. I haven't considered that.